### DISCUSSION TITLE
Query param with regex on List[str] param creates incorrect openapi doc

### DISCUSSION THREAD

**Discussion by ugtar** [fastapi/fastapi#8721] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional, List

from fastapi import FastAPI, Query


app = FastAPI()


@app.get("/")
async def root(
    field: Optional[List[str]] = Query(default=None, regex=r'^\w+:\w+$')
):
    return {"message": "Hello World"}
```

### Description

* open the browser to the auto-generated swagger docs
* expand the GET section for / and click "try it out"
![image](https://user-images.githubusercontent.com/96333/147686139-e29e49d4-15f0-41eb-9686-b4b4c9100fe6.png)
* add a single field matching the regex for example `foo:bar`. execute should work 
![image](https://user-images.githubusercontent.com/96333/147686235-b9b5a8d8-cb22-44a1-b02a-6cb27d259da5.png)
* try to add a second field also matching the validation regex, such as `bar:baz`, now swagger ui will not let you execute (the "add string item" button shakes and is red 
![image](https://user-images.githubusercontent.com/96333/147686327-9ec80e21-d1d9-4c72-94ff-742feac81226.png)

It appears the issue is due to the openapi doc that fast api is producing
```json
  "paths": {
    "/": {
      "get": {
        "summary": "Root",
        "operationId": "root__get",
        "parameters": [
          {
            "required": false,
            "schema": {
              "title": "Field",
              "pattern": "^\\w+:\\w+$",
              "type": "array",
              "items": {
                "pattern": "^\\w+:\\w+$",
                "type": "string"
              }
            },
            "name": "field",
            "in": "query"
          }
        ],
```
Specifically, the `pattern` is being applied to the schema of the `array` itself, not just the individual `items` in the array as would be expected.

I've confirmed this using swagger ui's cloud hosted api builder here: https://app.swaggerhub.com/apis/ugtar/foo/1.0.0
You can see that it demonstrates similar behavior (only one item is allowed by swagger ui).
![image](https://user-images.githubusercontent.com/96333/147686973-a4410965-adfd-4e68-9984-6bfaee82e583.png)

 However, if you remove or comment out the `pattern` on the array on line 34 of the openapi description, and just leave the `pattern` on the `items` subsection, you can see that now swagger ui allows you to send more than one `field` item
![image](https://user-images.githubusercontent.com/96333/147687092-9b68ea5a-ce2d-4647-9516-d55fe269634f.png)
and each of them is validated against the regex individually 
![image](https://user-images.githubusercontent.com/96333/147687141-d1e2f913-7110-452d-9473-338b9c658053.png)

### Operating System

Linux

### FastAPI Version

0.68.2

### Python Version

3.8.10

**Comment by ugtar** [comment]
I forgot to mention, importantly, the issue appears to be *only* with the openapi doc generation and thus affects the swagger ui. FastAPI itself (I tried the calls using curl) seems to validate the array field correctly item by item.

**Comment by harunyasar** [comment]
@ugtar I suspect this is not a valid use of regex. If you check [here](https://github.com/samuelcolvin/pydantic/blob/e8822775e3a331ca2d06a083429bf85840db6515/pydantic/fields.py#L284-L285), you will see that it only validates string type. Right before writing this comment, I was thinking that constraint type would be one of your solutions and after having seen [this issue](https://github.com/tiangolo/fastapi/issues/1021#issuecomment-590353181) I believe that you could simply use this solution.

```python
from pydantic import constr

QueryRegex = constr(regex="^\w+:\w+$")

@app.get("/")
async def root(field: Optional[List[QueryRegex]] = Query(None)):
    return {"message": "Hello World"}
```

**Reply by YuriiMotov** [reply, marked as answer]
More linter-friendly version:

```python
from typing import Annotated, Optional, TypeAlias

from fastapi import FastAPI, Query
from pydantic import Field

app = FastAPI()


MyType: TypeAlias = Annotated[str, Field(pattern=r"^\w+:\w+$")]


@app.get("/")
async def root(field: Annotated[Optional[list[MyType]], Query()] = None):
    return {"message": "Hello World"}

```

**Comment by ugtar** [comment]
@harunyasar I see what you mean, but I'm not sure that it's true. Pydantic seems to be applying the validator for strings to each element in a `List[str]`, which appears to be intentional behavior. I am a bit puzzled by that issue you mentioned as well since It is simple to verify that, If I use curl to get around the swagger issue, I can successfully send multiple fields, and they are properly validated against the regex. For example, if I send `curl -X GET /?field=foo:bar&field=bar:baz:` (note the trailing `:` on the second one which is not valid) it produces an error like this:
```json
{"detail":[{"loc":["query","field",1],"msg":"string does not match regex \"^\\w+:\\w+$\"","type":"value_error.str.regex","ctx":{"pattern":"^\\w+:\\w+$"}}]}
```
which even indicates the correct element (1) in the list that doesn't match. So it looks like pydantic is validating this correctly (at least as I would expect given the code as written). As I found, the only issue is with the generated openapi doc, not with the behavior of fastapi (or pydantic).

**Comment by harunyasar** [comment]
@ugtar I see what you mean. Apparently, I overlooked your other comment. Sorry about that.
